File: sorting/part2/mergesort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to merge two halves of the array
void merge(int arr[], int low, int mid, int high) {
    vector<int> temp;
    int left = low;
    int right = mid + 1;

    // Merging the two halves
    while (left <= mid && right <= high) {
        if (arr[left] <= arr[right]) {
            temp.push_back(arr[left]);
            left++;
        } else {
            temp.push_back(arr[right]);
            right++;
        }
    }

    // Adding remaining elements of left half
    while (left <= mid) {
        temp.push_back(arr[left]);
        left++;
    }

    // Adding remaining elements of right half
    while (right <= high) {
        temp.push_back(arr[right]);
        right++;
    }

    // Copying the sorted elements back to the original array
    for (int i = 0; i < temp.size(); i++) {
        arr[low + i] = temp[i];
    }
}

// Merge Sort function
void mergesort(int arr[], int low, int high) {
    if (low >= high) {
        return;
    }
    int mid = (low + high) / 2;
    mergesort(arr, low, mid);          // Recursively sorting the left half
    mergesort(arr, mid + 1, high);     // Recursively sorting the right half
    merge(arr, low, mid, high);       // Merging the two halves
}

int main() {
    int arr[] = {5, 2, 9, 1, 5, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    mergesort(arr, 0, n - 1);  // Calling mergesort

    cout << "Sorted array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}

File: sorting/part2/mergesort.py
'''this is algorithm for implementing merge sort
 this is a divide and conquer type of algo her we have to get a mid prtion of the array
 then recursively call it for the start fromt mid and mid+1 till end and we will continue this unil low is greater than high
 then we will use the merge funtion which will merge the array parts by storing it in temp array and then
 we will store it back to the actual array from the transfered index low'''
def merge(arr,low,mid,high):
    temp = []
    left = low
    right = mid+1
    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            temp.append(arr[left])
            left+=1
        else:
            temp.append(arr[right])
            right+=1
    while left <= mid:
        temp.append(arr[left])
        left+=1
    while right <= high:
        temp.append(arr[right])
        right+=1
    for i in range(len(temp)):
        arr[low+i] = temp[i]
def mergesort(arr,low,high):
    if low>=high:
        return
    mid = (high+low)//2
    mergesort(arr,low,mid)
    mergesort(arr,mid+1,high)
    merge(arr,low,mid,high)
arr = [64, 34, 25, 12, 22, 11, 90]
mergesort(arr, 0, len(arr) - 1)
print("Sorted array:", arr)